server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -0,0 +1,220 @@
+        this.joinHelper = new JoinHelper(settings, allocationService, masterService, transportService, this::getCurrentTerm,
+            this::handleJoinRequest);
+        this.persistedStateSupplier = persistedStateSupplier;
+        this.lastKnownLeader = Optional.empty();
+        this.lastJoin = Optional.empty();
+    }
+
+    private Optional<Join> ensureTermAtLeast(DiscoveryNode sourceNode, long targetTerm) {
+        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+        if (getCurrentTerm() < targetTerm) {
+            return Optional.of(joinLeaderInTerm(new StartJoinRequest(sourceNode, targetTerm)));
+        }
+        return Optional.empty();
+    }
+
+    private Join joinLeaderInTerm(StartJoinRequest startJoinRequest) {
+        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+        logger.debug("joinLeaderInTerm: from [{}] with term {}", startJoinRequest.getSourceNode(), startJoinRequest.getTerm());
+        Join join = coordinationState.get().handleStartJoin(startJoinRequest);
+        lastJoin = Optional.of(join);
+        if (mode != Mode.CANDIDATE) {
+            becomeCandidate("joinLeaderInTerm");
+        }
+        return join;
+    }
+
+    public void handleJoinRequest(JoinRequest joinRequest, JoinCallback joinCallback) {
+        assert Thread.holdsLock(mutex) == false;
+        transportService.connectToNode(joinRequest.getSourceNode());
+
+        synchronized (mutex) {
+            handleJoinRequestUnderLock(joinRequest, joinCallback);
+        }
+    }
+
+    private void handleJoinRequestUnderLock(JoinRequest joinRequest, JoinCallback joinCallback) {
+        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+        logger.trace("handleJoinRequestUnderLock: as {}, handling {}", mode, joinRequest);
+
+        final CoordinationState coordState = coordinationState.get();
+        final boolean prevElectionWon = coordState.electionWon();
+
+        if (joinRequest.getOptionalJoin().isPresent()) {
+            final Join join = joinRequest.getOptionalJoin().get();
+            // if someone thinks we should be master, let's add our vote and try to become one
+            // note that the following line should never throw an exception
+            ensureTermAtLeast(getLocalNode(), join.getTerm()).ifPresent(coordState::handleJoin);
+
+            // if we have already won the election, then the actual join does not matter for election purposes
+            if (coordState.electionWon()) {
+                // add join on a best-effort basis
+                try {
+                    coordState.handleJoin(join);
+                } catch (CoordinationStateRejectedException e) {
+                    logger.trace("failed to add join, ignoring", e);
+                }
+            } else {
+                coordState.handleJoin(join); // this might fail and bubble up the exception
+            }
+        }
+
+        switch (mode) {
```

Comment:
Great. I now wonder if this whole block can move into the `JoinHelper`. If the code above returned its `mode` and called `becomeLeader()` then the `JoinHelper` should be able to work out the right things to do with the join.

Reply:
I've tried, but that would make JoinHelper be aware of the mode, as well as need to call `becomeLeader`, so it would be calling back into Coordinator (or CoordinationState), which also feels wrong. Finally, these lines here need to happen under the mutex, so even if I move part of this over to Coordinator, it can not just selectively call into Coordinator.

Reply:
I've pushed my attempt at this. I'm not sure if it's an improvement over the existing code though.

Reply:
attempt is here: 6f58a8c1c57